File: j.cpp
#include <bits/stdc++.h>
using namespace std;

int b_pref[100001];
int a[100001], b[100001];

int sum_b(int start, int end)
{
    if(end < start) return 0;
    if(start > 0) return sum_b(0, end) - sum_b(0, start - 1);
    return b_pref[end];
}

int main()
{
    auto in = ifstream("rapunzel.in");

    int T;
    in >> T;
    while(T--)
    {
        int N, K, L;
        in >> N >> K >> L;
        for(int i = 0; i < N; i++) in >> a[i];
        for(int i = 0; i < N; i++) in >> b[i];

        b_pref[0] = b[0];
        for(int i = 1; i < N; i++)
            b_pref[i] = b_pref[i-1] + b[i];
        
        int best = -1;

        for(int i = 0; i < N; i++)
        {
            if(best != -1) break;

            int lo = i, hi = N;
            while(lo < hi)
            {
                int mid = (lo + hi) / 2;
                int x = L * a[i] + sum_b(i+1, mid);

                if(x >= K - 1 && x <= K + 1)
                {
                    best = i+1;
                    break;
                }

                if(x > K+1) hi = mid;
                else if(x < K-1) lo = mid+1;
            }
        }

        cout << best << endl;
    }
}

File: j.py
f = open('rapunzel.in')

T = int(f.readline().strip())

b_pref = []

def sum_b(start, end):
    global b_pref
    if end < start: return 0
    if start > 0:
        return sum_b(0, end) - sum_b(0, start - 1)
    return b_pref[end]

b_pref = [0 for _ in range(100001)]
for _ in range(T):
    n, k, l = map(int, f.readline().strip().split())
    a = list(map(int, f.readline().strip().split()))
    b = list(map(int, f.readline().strip().split()))

    b_pref[0] = b[0]
    for i in range(1, len(b)):
        b_pref[i] = b_pref[i-1] + b[i]
    
    best = -1
    for i in range(n):
        if best != -1: break

        lo = i
        hi = n

        while abs(hi - lo) > 2:
            mid = (lo + hi) // 2

            x = l * a[i] + sum_b(i+1, mid)

            #print(i, mid, x)

            if x >= k - 1 and x <= k + 1:
                best = i+1
                break
            
            if x > k + 1:
                hi = mid
            elif x < k - 1:
                lo = mid + 1
        
        if best == -1:
            for v in range(lo, hi):
                x = l * a[i] + sum_b(i+1, v)
                if x >= k - 1 and x <= k + 1:
                    best = i+1

    print(best)
